### PR TITLE
Sets matplotlib backend to agg before rendering math

### PR DESCRIPTION
**Pull request by abidlabs**
This PR:

* sets the `matplotlib` backend to `Agg` before trying to render LaTeX (this is required for LaTeX to render properly)
* warns a user if `matplotlib` backend is not `Agg` when using the `gr.Plot` component with a `matplotlib` plot 

Closes: #4018, Closes: #3943 (hopefully, but needs to be tested on a Mac)

### CODE REVIEW

File: gradio/utils.py [hunk 1]
```diff
@@ -900,34 +899,45 @@ def __str__(self):
...
+        height_match = re.search(r'height="([\d.]+)pt"', svg_code)
+        if height_match:
+            height = float(height_match.group(1))
+            new_height = height / FONTSIZE  # conversion from pt to em
+            svg_code = re.sub(
+                r'height="[\d.]+pt"', f'height="{new_height}em"', svg_code
+            )
+        copy_code = f"<span style='font-size: 0px'>{formula}</span>"
+        return f"{copy_code}{svg_code}"
+    finally:
```
Comment (abidlabs) [inline gradio/utils.py:939]: Note: `finally` blocks get executed before the `return` above even if the `try` block runs successfully

File: gradio/components.py [hunk 2]
```diff
@@ -4856,6 +4856,10 @@ def postprocess(self, y) -> Dict[str, str] | None:
         if y is None:
             return None
         if isinstance(y, (ModuleType, matplotlib.figure.Figure)):  # type: ignore
+            if not matplotlib.get_backend().lower() == "agg":
+                warnings.warn(
+                    "Matplotlib backend is not set to 'Agg'. Plot may not display correctly. To fix this, add the following to the top of your script: import matplotlib; matplotlib.use('Agg')"
```
Comment (freddyaboulton) [inline gradio/components.py:4861]: I wonder if we can raise this somewhere up the chain (plot component init?). May be some false warnings but I worry the app would have already crashed before this is printed.

**Review by freddyaboulton — APPROVED** [review]
Can confirm this fixes the original problem! Code looks good - only one comment. Thanks @abidlabs !

File: gradio/components.py (hunk 2 above)
Comment (abidlabs) [reply gradio/components.py:4861]: Good point, let me think about this some more as we don't want too many false warnings.

Comment (abidlabs) [reply gradio/components.py:4861]: What if we were to try/catch the function execution, specifically looking for "main thread is not in main loop"

Comment (freddyaboulton) [reply gradio/components.py:4861]: Yea I think that will work

File: gradio/utils.py [hunk 3]
```diff
@@ -899,34 +899,48 @@ def __str__(self):
         return ""
 
 
+class MatplotlibBackendMananger:
```
Comment (freddyaboulton) [inline gradio/utils.py:902]: Nice!

**Comment by abidlabs** [comment]
Refactored the matplotlib logic into a context manager so now we can use it everywhere. This has the advantage of users not needing to put `matplotlib.use("Agg")` at the top of their code. Most of the additional changes are just removing these extraneous lines from our demos.

@freddyaboulton would you be able to give this another pass?

**Review by freddyaboulton — APPROVED** [review]
There's a call to `process_api` in the Blocks `__call__` method but maybe that's ok since it's running in the parent thread.

Anyways, this LGTM! Thanks @abidlabs !

**Comment by abidlabs** [comment]
> There's a call to process_api in the Blocks __call__ method but maybe that's ok since it's running in the parent thread.

Yes, I think that one should work as is!

Thanks for reviewing! Will take a look at these tests and then merge in.

**Comment by abidlabs** [comment]
Strange, all of these tests are passing for me locally on Windows...

**Comment by abidlabs** [comment]
Tests are passing, will merge. Thanks once again @freddyaboulton for the review!

**Pull request merged**